src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -0,0 +1,563 @@
+
+            return quicStream.HandleEvent(ref connectionEventStruct);
+        }
+
+        private uint HandleEvent(ref StreamEvent evt)
+        {
+            uint status = MsQuicConstants.Success;
+
+            try
+            {
+                switch (evt.Type)
+                {
+                    case QUIC_STREAM_EVENT.START_COMPLETE:
+                        status = HandleStartComplete();
+                        break;
+                    case QUIC_STREAM_EVENT.RECV:
+                        {
+                            HandleEventRecv(ref evt);
+                        }
+                        break;
+                    case QUIC_STREAM_EVENT.SEND_COMPLETE:
+                        {
+                            status = HandleEventSendComplete(ref evt);
+                        }
+                        break;
+                    case QUIC_STREAM_EVENT.PEER_SEND_CLOSE:
+                        {
+                            status = HandleEventPeerSendClose();
+                        }
+                        break;
+                    // TODO figure out difference between SEND_ABORT and RECEIVE_ABORT
+                    case QUIC_STREAM_EVENT.PEER_SEND_ABORT:
+                        {
+                            // TODO why is this firing.
+                            status = HandleEventPeerSendAbort();
+                        }
+                        break;
+                    case QUIC_STREAM_EVENT.PEER_RECV_ABORT:
+                        {
+                            status = HandleEventPeerRecvAbort();
+                        }
+                        break;
+                    case QUIC_STREAM_EVENT.SEND_SHUTDOWN_COMPLETE:
+                        {
+                            status = HandleEventSendShutdownComplete(ref evt);
+                        }
+                        break;
+                    case QUIC_STREAM_EVENT.SHUTDOWN_COMPLETE:
+                        {
+                            // TODO make sure close is called.
+                            status = HandleEventShutdownComplete();
+                        }
+                        break;
+                    default:
+                        break;
+                }
+            }
+            catch (Exception)
+            {
+                return MsQuicConstants.InternalError;
+            }
+
+            return status;
+        }
+
+        private void HandleEventRecv(ref MsQuicNativeMethods.StreamEvent evt)
+        {
+            static unsafe void CopyToBuffer(Span<byte> buffer, StreamEvent evt)
+            {
+                int length = (int)evt.Data.Recv.Buffers[0].Length;
+                new Span<byte>(evt.Data.Recv.Buffers[0].Buffer, length).CopyTo(buffer);
+            }
+
+            // Need to think hard about backpressure here.
+            // Can this pipe grow infinitely if the consumer isn't reading large chunks?
+            // TODO add a test which does a bunch of large writes and small reads.
```

Comment:
This is going to be tricky to get right with your design. MsQuic dynamically grows receive buffers based on the rate at which the app can process the data. So if you're constantly just copying somewhere and returning, and don't tell us to stop, then the buffers are going to get pretty big.

Reply:
I don't like this extra copy in general. Can we instead hold onto the buffer indefinitely and call `StreamReceiveComplete` in chunks as we work our way through it?

Reply:
Yes. But you can only call StreamReceiveComplete once per RECEIVE event.

Reply:
> This is going to be tricky to get right with your design. MsQuic dynamically grows receive buffers based on the rate at which the app can process the data. So if you're constantly just copying somewhere and returning, and don't tell us to stop, then the buffers are going to get pretty big.

So, I purposely set the limits on the pipe to make it so backpressure is always hit when doing a write if no one is there to read it. This means we would be telling MsQuic to stop by calling DisableReceive. This effectively makes the pipe a single buffer at all points. 

What I'm more worried about is if the writer is outpacing the reader (which I'll need to talk to @davidfowl about). Effective what I think could happen is if the calls to ReadAsync are small (ex reading a single buffer at a time) and the calls to WriteAsync are large, the pipe can continue to grow in size. I'll chat with him offline about it.

> I don't like this extra copy in general. Can we instead hold onto the buffer indefinitely and call StreamReceiveComplete in chunks as we work our way through it?

To make sure I understand, @nibanks is it safe to hold onto the QuicBuffers until ReceiveComplete is called? If so, that could save us the copy always. However, it does introduce some complexity by reimplementing completion, cancellation, and read synchronization that pipes gives us for free.

Reply:
> This is going to be tricky to get right with your design. MsQuic dynamically grows receive buffers based on the rate at which the app can process the data. So if you're constantly just copying somewhere and returning, and don't tell us to stop, then the buffers are going to get pretty big.

So, I purposely set the limits on the pipe to make it so backpressure is always hit when doing a write if no one is there to read it. This means we would be telling MsQuic to stop by calling DisableReceive. This effectively makes the pipe a single buffer at all points. 

What I'm more worried about is if the writer is outpacing the reader (which I'll need to talk to @davidfowl about). Effective what I think could happen is if the calls to ReadAsync are small (ex reading a single buffer at a time) and the calls to WriteAsync are large, the pipe can continue to grow in size. I'll chat with him offline about it.

> I don't like this extra copy in general. Can we instead hold onto the buffer indefinitely and call StreamReceiveComplete in chunks as we work our way through it?

To make sure I understand, @nibanks is it safe to hold onto the QuicBuffers until ReceiveComplete is called? If so, that could save us the copy always. However, it does introduce some complexity by reimplementing completion, cancellation, and read synchronization that pipes gives us for free.

Reply:
> To make sure I understand, @nibanks is it safe to hold onto the QuicBuffers until ReceiveComplete is called? If so, that could save us the copy always. However, it does introduce some complexity by reimplementing completion, cancellation, and read synchronization that pipes gives us for free.

To clarify: I'd like us to grab the buffer via the event, return from the event, and be able to `StreamReceiveComplete` at a later time (once the user calls `Read` enough to exhaust that buffer), maybe in a different thread. I think this would make it relatively simple.

Reply:
You can do that, but you must return QUIC_STATUS_PENDING from the event.

Reply:
I'm a bit confused by what the difference between returning QUIC_STATUS_PENDING vs ReceiveComplete(0) is here then.

> I think this would make it relatively simple.

Famous last words 😄 .